cuda/base/kernel_launch_reduction.cuh

```diff
@@ -0,0 +1,529 @@
+            static_cast<int64>(size), fn, op,
+            [] __device__(auto v) { return v; }, as_cuda_type(init),
+            as_cuda_type(partial.get_data()), map_to_device(args)...);
+        generic_kernel_reduction_1d<<<1, block_size>>>(
+            static_cast<int64>(num_blocks),
+            [] __device__(auto i, auto v) { return v[i]; }, op, finalize,
+            as_cuda_type(init), as_cuda_type(result),
+            as_cuda_type(partial.get_const_data()));
+    } else {
+        generic_kernel_reduction_1d<<<1, block_size>>>(
+            static_cast<int64>(size), fn, op, finalize, as_cuda_type(init),
+            as_cuda_type(result), map_to_device(args)...);
+    }
+}
+
+
+template <typename ValueType, typename KernelFunction, typename ReductionOp,
+          typename FinalizeOp, typename... KernelArgs>
+void run_kernel_reduction(std::shared_ptr<const CudaExecutor> exec,
+                          KernelFunction fn, ReductionOp op,
+                          FinalizeOp finalize, ValueType init,
+                          ValueType* result, dim<2> size, KernelArgs&&... args)
+{
+    constexpr int oversubscription = 16;
+    gko::cuda::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = default_block_size;
+    const auto rows = static_cast<int64>(size[0]);
+    const auto cols = static_cast<int64>(size[1]);
+    const auto num_blocks =
+        std::min<int64>(ceildiv(rows * cols, block_size),
+                        exec->get_num_warps() * oversubscription);
+    if (num_blocks > 1) {
+        Array<ValueType> partial{exec, static_cast<size_type>(num_blocks)};
+        generic_kernel_reduction_2d<<<num_blocks, block_size>>>(
+            rows, cols, fn, op, [] __device__(auto v) { return v; },
+            as_cuda_type(init), as_cuda_type(partial.get_data()),
+            map_to_device(args)...);
+        generic_kernel_reduction_1d<<<1, block_size>>>(
+            static_cast<int64>(num_blocks),
+            [] __device__(auto i, auto v) { return v[i]; }, op, finalize,
+            as_cuda_type(init), as_cuda_type(result),
+            as_cuda_type(partial.get_const_data()));
+    } else {
+        generic_kernel_reduction_2d<<<1, block_size>>>(
+            rows, cols, fn, op, finalize, as_cuda_type(init),
+            as_cuda_type(result), map_to_device(args)...);
+    }
+}
+
+
+template <int subwarp_size, typename ValueType, typename KernelFunction,
+          typename ReductionOp, typename FinalizeOp, typename... KernelArgs>
+__global__
+    __launch_bounds__(default_block_size) void generic_kernel_row_reduction_2d(
+        int64 rows, int64 cols, int64 col_blocks, KernelFunction fn,
+        ReductionOp op, FinalizeOp finalize, ValueType init, ValueType* result,
+        int64 result_stride, KernelArgs... args)
+{
+    const auto idx = thread::get_subwarp_id_flat<subwarp_size, int64>();
+    const auto row = idx % rows;
+    const auto col_block = idx / rows;
```

Comment:
Does it use column-major?

Reply:
I'm not sure if that term fits well here - we use one subwarp per row, so the reads are usually coalescing if the input is ready from a row-major matrix.

Reply:
ah, I see. It is the subwarp id, not the thread id.